synapseclient/api/entity_services.py

```diff
@@ -0,0 +1,173 @@
+if TYPE_CHECKING:
+    from synapseclient import Synapse
+
+
+async def post_entity(
+    request: Dict[str, Any],
+    generated_by: Optional[str] = None,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, Any]:
+    """
+    Arguments:
+        request: The request for the entity matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/Entity.html>
+        generated_by: The ID of the activity to associate with the entity.
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+    Returns:
+        The requested entity matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/Entity.html>
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+    params = {}
+    if generated_by:
+        params["generatedBy"] = generated_by
+    return await client.rest_post_async(
+        uri="/entity", body=json.dumps(request), params=params
+    )
+
+
+async def put_entity(
+    entity_id: str,
+    request: Dict[str, Any],
+    new_version: bool = False,
+    generated_by: Optional[str] = None,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, Any]:
+    """
+    Arguments:
+        entity_id: The ID of the entity to update.
+        request: The request for the entity matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/Entity.html>
+        new_version: If true, a new version of the entity will be created.
+        generated_by: The ID of the activity to associate with the entity.
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+    Returns:
+        The requested entity bundle matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/Entity.html>
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+    params = {}
+    if generated_by:
+        params["generatedBy"] = generated_by
+    if new_version:
+        params["newVersion"] = "true"
+    return await client.rest_put_async(
+        uri=f"/entity/{entity_id}", body=json.dumps(request), params=params
+    )
+
+
+async def get_entity(
+    entity_id: str,
+    synapse_client: Optional["Synapse"] = None,
+) -> Dict[str, Any]:
+    """
+    Arguments:
+        entity_id: The ID of the entity.
+        synapse_client: If not passed in or None this will use the last client from
+            the `.login()` method.
+
+    Returns:
+        The requested entity bundle matching
+            <https://rest-docs.synapse.org/rest/org/sagebionetworks/repo/model/Entity.html>
+    """
+    from synapseclient import Synapse
+
+    client = Synapse.get_client(synapse_client=synapse_client)
+    return await client.rest_get_async(
+        uri=f"/entity/{entity_id}",
+    )
+
+
+@alru_cache(ttl=60)
```

Comment:
This was a nice performance gain when looking at cases where multiple files are getting stored in a container. This package works in such a way that if multiple things call this co-routine it is only executed the FIRST time it is called - The rest of the calls await the result.

In the 1000 Files, 1 GiB total size test case this had around a 25% time improvement.

https://github.com/aio-libs/async-lru

Reply:
Just out of curiosity, what's the issue when using the built in lru_cache?

Reply:
The problem with lru_cache is the way that it works doesn't work for asyncio. What I mean by this:

Suppose we have 10 coroutines running to save 10 files. Each of those 10 files needs to determine where the file is going to be uploaded and calls this `get_upload_destination()` method. lru_cache works in this way:

1. If a result is already present in the cache return the result to the caller
2. If a result is not present in the cache enter the function and cache the result
In the case of asyncio - all 10 coroutines are going to enter this `get_upload_destination` because the result hasn't yet returned and been cached.

ALRU cache works by:

1. If a result is already present in the cache return the result to the caller
2. If a result is not present in the cache and a future already exists that promises the result - return that future to the caller (This means all 10 coroutines will wait on a single Future to get the result)
3. If a result is not present and no future is already retrieving the result - enter the function, create a future that promises the result, execute, and cache the result.